File: Direct/iOS/SampleApp/SampleAppObjc/SampleAppObjc/MainViewHandler.m
//
//  MainViewHandler.m
//  SampleAppObjc
//
//  Created by Maria on 31/07/2023.
//

#import "MainViewHandler.h"

@interface MainViewHandler()
@property (nonatomic, retain) BaseMainViewController* baseMainViewController;
@property (nonatomic, retain) PradoBannerView* bannerView;
@property (nonatomic, retain) PradoInterstitialAd* interstitialAd;
@property (nonatomic, retain) PradoRewardedAd* rewardedAd;
@end

@implementation MainViewHandler

@synthesize baseMainViewController;
@synthesize bannerView;
@synthesize interstitialAd;
@synthesize rewardedAd;

- (void)setViewControllerWithBaseMainViewController:(BaseMainViewController *)baseMainViewController {
    self.baseMainViewController = baseMainViewController;
}

- (NSString * _Nonnull)getAppLabel {
    return @"Prado Sample App";
}

- (NSString * _Nonnull)getSDKVersion {
    return [Prado getSdkVersion];
}

- (void)initSDK {
    [Prado.instance initializeWithPublisherID:@"14428" securityToken:@"6yAsKUngaG5yC4X5HsRoatKTso40NMoZ" delegate:self];
}

- (void)onInitSuccess {
    [self initBannerWithView];
    [baseMainViewController onSDKInitSuccess];
}

- (void)onInitError:(NSString*) errorMessage {
    [baseMainViewController onSDKInitFailureWithError: errorMessage];
}

// MARK: - Banner
- (void)loadBanner {
    if ([Prado.instance isSDKInitialized]) {
        [bannerView load];
    } else {
        [baseMainViewController logOutWithMessage:@"SDK not initialized"];
    }
}

- (void)closeBanner {
    [bannerView close];
    [bannerView removeFromSuperview];
    bannerView = NULL;
    [self initBannerWithView];
}

- (void)onBannerAdLoadedWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView {
    [baseMainViewController onBannerLoaded];
}

- (void)onBannerAdFailedToLoadWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView error:(PradoError * _Nonnull)error {
    [baseMainViewController onBannerLoadFailedWithError: error.description];
}

- (void)onBannerAdShownWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView {
    [baseMainViewController onBannerOpened];
}

- (void)onBannerAdFailedToShowWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView error:(PradoError * _Nonnull)error {
    [baseMainViewController onBannerShowFailedWithError: error.description];
}

- (void)onBannerAdImpressionWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView {
    [baseMainViewController onBannerImpression];
}

- (void)onBannerAdClosedWithPradoBannerView:(PradoBannerView * _Nonnull)pradoBannerView {
    [baseMainViewController onBannerClosed];
}

// MARK: - Interstitial ad
- (void)loadInterstitial {
    if ([Prado.instance isSDKInitialized]) {
        [PradoInterstitialAd loadWithDelegate: self];
    } else {
        [baseMainViewController logOutWithMessage: @"SDK not initialized"];
    }
}

- (void)showInterstitial {
    if (interstitialAd != NULL && [interstitialAd isLoaded] ) {
        [interstitialAd showWithViewController: baseMainViewController];
    } else {
        [baseMainViewController logOutWithMessage: @"Interstitial not ready"];
    }
}

- (void)onInterstitialAdLoadedWithPradoInterstitialAd:(PradoInterstitialAd * _Nonnull)pradoInterstitialAd {
    self.interstitialAd = pradoInterstitialAd;
    [baseMainViewController onInterstitialLoaded];
}

- (void)onInterstitialAdFailedToLoadWithError:(PradoError *)error {
    [baseMainViewController onInterstitialLoadFailedWithError: error.description];
}

- (void)onInterstitialAdFailedToShowWithPradoInterstitialAd:(PradoInterstitialAd * _Nonnull)pradoInterstitialAd pradoError:(PradoError * _Nonnull)pradoError {
    [baseMainViewController onInterstitialShowFailedWithError: pradoError.description];
}

- (void)onInterstitialAdShownWithPradoInterstitialAd:(PradoInterstitialAd * _Nonnull)pradoInterstitialAd {
    [baseMainViewController onInterstitialOpened];
}

- (void)onInterstitialImpressionWithPradoInterstitialAd:(PradoInterstitialAd * _Nonnull)pradoInterstitialAd {
    [baseMainViewController onInterstitialImpression];
}

- (void)onInterstitialAdClosedWithPradoInterstitialAd:(PradoInterstitialAd * _Nonnull)pradoInterstitialAd {
    [baseMainViewController onInterstitialClosed];
}

// MARK: - Rewarded ad
- (void)loadRewarded {
    if ([Prado.instance isSDKInitialized]) {
        [PradoRewardedAd loadWithDelegate: self];
    } else {
        [baseMainViewController logOutWithMessage: @"SDK not initialized"];
    }
}

- (void)showRewarded {
    if (rewardedAd != NULL && [rewardedAd isLoaded] ) {
        [rewardedAd showWithViewController: baseMainViewController];
    } else {
        [baseMainViewController logOutWithMessage: @"Rewarded not ready"];
    }
}

- (void)onRewardedAdLoadedWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd {
    rewardedAd = pradoRewardedAd;
    [baseMainViewController onRewardedLoaded];
}

- (void)onRewardedAdFailedToLoadWithPradoError:(PradoError * _Nonnull)pradoError {
    [baseMainViewController onRewardedLoadFailedWithError: pradoError.description];
}

- (void)onRewardedAdShownWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd {
    [baseMainViewController onRewardedOpened];
}

- (void)onRewardedAdFailedToShowWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd pradoError:(PradoError * _Nonnull)pradoError {
    [baseMainViewController onRewardedShowFailedWithError: pradoError.description];
}

- (void)onRewardedImpressionWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd {
    [baseMainViewController onRewardedImpression];
}

- (void)onRewardReceivedWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd {
    [baseMainViewController onRewardAchieved];
}

- (void)onRewardedAdClosedWithPradoRewardedAd:(PradoRewardedAd * _Nonnull)pradoRewardedAd {
    [baseMainViewController onRewardedClosed];
}

-(void) initBannerWithView {
    if ([baseMainViewController view] != NULL) {
        bannerView = [[PradoBannerView alloc] init];
        bannerView.delegate = self;
        [bannerView setTranslatesAutoresizingMaskIntoConstraints: NO];
        [[baseMainViewController view]addSubview: bannerView];

        NSLayoutConstraint *bottom = [
            NSLayoutConstraint constraintWithItem:bannerView
            attribute:NSLayoutAttributeBottom
            relatedBy:NSLayoutRelationEqual
            toItem:baseMainViewController.bottomLayoutGuide
            attribute:NSLayoutAttributeTop
            multiplier:1
            constant:0
        ];

        NSLayoutConstraint *centerX = [
            NSLayoutConstraint constraintWithItem:bannerView
            attribute:NSLayoutAttributeCenterX
            relatedBy:NSLayoutRelationEqual
            toItem:baseMainViewController.view
            attribute:NSLayoutAttributeCenterX
            multiplier:1
            constant:0
        ];

        NSLayoutConstraint *height = [
            NSLayoutConstraint constraintWithItem:bannerView
            attribute:NSLayoutAttributeHeight
            relatedBy:NSLayoutRelationEqual
            toItem:NULL
            attribute:NSLayoutAttributeNotAnAttribute
            multiplier:0
            constant: 50
        ];

        NSLayoutConstraint *width = [
            NSLayoutConstraint constraintWithItem:bannerView
            attribute:NSLayoutAttributeWidth
            relatedBy:NSLayoutRelationEqual
            toItem:NULL
            attribute:NSLayoutAttributeNotAnAttribute
            multiplier:0
            constant: 320
        ];

        [[baseMainViewController view] addConstraints:@[bottom, centerX, height, width]];
    }
}

@end

File: Prado Direct/iOS/SampleApp/SampleAppSwift/SampleAppSwift/MainViewHandler.swift
//
//  MainViewHandler.swift
//  SampleApp
//
//  Created by Maria on 25/07/2023.
//

import Foundation
import PradoSDK
import CommonSampleApp

public class MainViewHandler: NSObject, MainViewControllerHandler, PradoInitDelegate, PradoInterstitialDelegate, PradoRewardedDelegate, PradoBannerDelegate {
    
    var baseMainViewController: BaseMainViewController!
    var bannerView: PradoBannerView!
    var interstitialAd: PradoSDK.PradoInterstitialAd?
    var rewardedAd: PradoSDK.PradoRewardedAd?
    
    public func setViewController(baseMainViewController: BaseMainViewController) {
        self.baseMainViewController = baseMainViewController
    }
    
    public func getAppLabel() -> String {
        return "Prado Sample App"
    }
    
    public func getSDKVersion() -> String {
        return Prado.getSdkVersion()
    }
    
    public func initSDK() {
        Prado.instance.initialize(publisherID: "15492", securityToken: "4b4982c7c7429ba2006852c2ef9f684f", delegate: self)
    }
    
    public func onInitSuccess() {
        initBannerWithView()
        baseMainViewController.onSDKInitSuccess()
    }
    
    public func onInitError(_ errorMessage: String) {
        baseMainViewController.onSDKInitFailure(error: errorMessage)
    }
    
    // MARK: - Banner
    public func loadBanner() {
        if Prado.instance.isSDKInitialized() {
            bannerView.load()
        } else {
            baseMainViewController.logOut(message: "SDK not initialized")
        }
    }
    
    public func closeBanner() {
        bannerView.close()
        bannerView.removeFromSuperview()
        bannerView = nil
        initBannerWithView()
    }
    
    public func onBannerAdLoaded(pradoBannerView: PradoSDK.PradoBannerView) {
        baseMainViewController.onBannerLoaded()
    }
    
    public func onBannerAdFailedToLoad(pradoBannerView: PradoSDK.PradoBannerView, error: PradoSDK.PradoError) {
        baseMainViewController.onBannerLoadFailed(error: error.description)
    }
    
    public func onBannerAdShown(pradoBannerView: PradoSDK.PradoBannerView) {
        baseMainViewController.onBannerOpened()
    }
    
    public func onBannerAdFailedToShow(pradoBannerView: PradoSDK.PradoBannerView, error: PradoSDK.PradoError) {
        baseMainViewController.onBannerShowFailed(error: error.description)
    }
    
    public func onBannerAdImpression(pradoBannerView: PradoSDK.PradoBannerView) {
        baseMainViewController.onBannerImpression()
    }
    
    public func onBannerAdClosed(pradoBannerView: PradoSDK.PradoBannerView) {
        baseMainViewController.onBannerClosed()
    }

    // MARK: - Interstitial ad
    public func loadInterstitial() {
        if Prado.instance.isSDKInitialized() {
            PradoInterstitialAd.load(delegate: self)
        } else {
            baseMainViewController.logOut(message: "SDK not initialized")
        }
    }
    
    public func showInterstitial() {
        if let intestitial = interstitialAd, intestitial.isLoaded() {
            intestitial.show(viewController: baseMainViewController)
        } else {
            baseMainViewController.logOut(message: "Interstitial not ready")
        }
    }
    
    public func onInterstitialAdLoaded(pradoInterstitialAd ad: PradoSDK.PradoInterstitialAd) {
        self.interstitialAd = ad
        baseMainViewController.onInterstitialLoaded()
    }
    
    public func onInterstitialAdFailedToLoad(pradoError error: PradoSDK.PradoError) {
        baseMainViewController.onInterstitialLoadFailed(error: error.description)
    }
    
    public func onInterstitialAdFailedToShow(pradoInterstitialAd: PradoSDK.PradoInterstitialAd, pradoError: PradoSDK.PradoError) {
        baseMainViewController.onInterstitialShowFailed(error: pradoError.description)
    }
    
    public func onInterstitialAdShown(pradoInterstitialAd ad: PradoSDK.PradoInterstitialAd) {
        baseMainViewController.onInterstitialOpened()
    }
    
    public func onInterstitialImpression(pradoInterstitialAd: PradoSDK.PradoInterstitialAd) {
        baseMainViewController.onInterstitialImpression()
    }
    
    public func onInterstitialAdClosed(pradoInterstitialAd ad: PradoSDK.PradoInterstitialAd) {
        baseMainViewController.onInterstitialClosed()
    }
    
    // MARK: - Rewarded ad
    public func loadRewarded() {
        if Prado.instance.isSDKInitialized() {
            PradoRewardedAd.load(delegate: self)
        } else {
            baseMainViewController.logOut(message: "SDK not initialized")
        }
    }
    
    public func showRewarded() {
        if let rewarded = rewardedAd, rewarded.isLoaded() {
            rewarded.show(viewController: baseMainViewController)
        } else {
            baseMainViewController.logOut(message: "Rewarded not ready")
        }
    }
    
    public func onRewardedAdLoaded(pradoRewardedAd ad: PradoSDK.PradoRewardedAd) {
        rewardedAd = ad
        baseMainViewController.onRewardedLoaded()
    }
    
    public func onRewardedAdFailedToLoad(pradoError error: PradoSDK.PradoError) {
        baseMainViewController.onRewardedLoadFailed(error: error.description)
    }
    
    public func onRewardedAdShown(pradoRewardedAd ad: PradoSDK.PradoRewardedAd) {
        baseMainViewController.onRewardedOpened()
    }
    
    public func onRewardedAdFailedToShow(pradoRewardedAd: PradoSDK.PradoRewardedAd, pradoError: PradoSDK.PradoError) {
        baseMainViewController.onRewardedShowFailed(error: pradoError.description)
    }
    
    public func onRewardedImpression(pradoRewardedAd: PradoSDK.PradoRewardedAd) {
        baseMainViewController.onRewardedImpression()
    }
    
    public func onRewardReceived(pradoRewardedAd ad: PradoSDK.PradoRewardedAd) {
        baseMainViewController.onRewardAchieved()
    }
    
    public func onRewardedAdClosed(pradoRewardedAd ad: PradoSDK.PradoRewardedAd) {
        baseMainViewController.onRewardedClosed()
    }
    
    func initBannerWithView() {
        if let view = baseMainViewController.view {
            bannerView = PradoBannerView()
            bannerView.delegate = self
            bannerView.translatesAutoresizingMaskIntoConstraints = false
            view.addSubview(bannerView)
            view.addConstraints(
                [NSLayoutConstraint(
                    item: bannerView!,
                    attribute: .bottom,
                    relatedBy: .equal,
                    toItem: baseMainViewController.bottomLayoutGuide,
                    attribute: .top,
                    multiplier: 1,
                    constant: 0),
                 NSLayoutConstraint(
                    item: bannerView!,
                    attribute: .centerX,
                    relatedBy: .equal,
                    toItem: view,
                    attribute: .centerX,
                    multiplier: 1,
                    constant: 0),
                 NSLayoutConstraint(
                    item: bannerView!,
                    attribute: .height,
                    relatedBy: .equal,
                    toItem: nil,
                    attribute: .notAnAttribute,
                    multiplier: 0,
                    constant: 50),
                 NSLayoutConstraint(
                    item: bannerView!,
                    attribute: .width,
                    relatedBy: .equal,
                    toItem: nil,
                    attribute: .notAnAttribute,
                    multiplier: 0,
                    constant: 320)])

        }
    }
}
